### PR TITLE
[Bugfix] Fix misclassification of prefill as uniform decode for hybrid models

### PR DESCRIPTION
**Pull request by xiatwhu**
## Purpose
Fix misclassification of prefill as uniform decode for hybrid models

In Hybrid models, prefill and decode/spec_decode are handled with different logic. If a prefill is misidentified as spec_decode, the model will produce garbage output. When handling prefill, the state must be zeroed out (or not) based on has_initial_state; if this step is skipped, a random (uninitialized) state will be used in the computation.

https://github.com/vllm-project/vllm/blob/42365140980d2f56149822136e89d8365dfc80ca/vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py#L1503-L1514

https://github.com/vllm-project/vllm/blob/42365140980d2f56149822136e89d8365dfc80ca/vllm/model_executor/layers/mamba/mamba_mixer2.py#L832-L844

If the number of tokens in a request's prompt happens to equal exactly 1 + num_spec_tokens, that request's prefill computation will be mistakenly treated as spec_decode, thereby using an uninitialized state and producing garbage output.

https://github.com/vllm-project/vllm/blob/97b5ce5c3931173b3efebca58e597b8e58730cd3/vllm/v1/worker/gpu_model_runner.py#L3810-L3828

## Test Plan

```python
import vllm
from vllm import LLM, SamplingParams

# Number of tokens is 4, equal to (1 + num_speculative_tokens)
prompts = ["The capital of France"]                                                                                                                                                                                                    

sampling_params = SamplingParams(temperature=0, top_p=1, max_tokens=12)

if __name__ == '__main__':
    llm = LLM(model="Qwen/Qwen3.5-35B-A3B-GPTQ-Int4",
        max_model_len=1024,
        speculative_config = {"method":"qwen3_5_mtp","num_speculative_tokens":3},
    )

    for i in range(4):
        outputs = llm.generate(prompts, sampling_params)
        print(f"Input: {outputs[0].prompt!r}\nOutput: {outputs[0].outputs[0].text!r}\n")
```

## Test Result
Old Result
```
Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: 'калибровочный\n=  =\n\n=== Мор'

Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ' is a 2019 American animated science fiction comedy'

Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ' is a comprehensive guide to the history of the English language,'

Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ', the only viable path is to acknowledge the impossibility of'
```

New Result
```
Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ' is Paris.\nThe capital of France is Paris.\n'

Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ' is Paris.\nThe capital of France is Paris.\n'

Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ' is Paris.\nThe capital of France is Paris.\n'

Rendering prompts: 100%
Processed prompts: 100%
Input: 'The capital of France'
Output: ' is Paris.\nThe capital of France is Paris.\n'
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -4160,6 +4164,12 @@ def execute_model(
                     scheduler_output.num_common_prefix_blocks,
                 )
 
+            force_uniform_decode = None
+            if self.model_config.is_hybrid:
+                iteration_details = compute_iteration_details(scheduler_output)
+                if iteration_details.num_ctx_requests > 0:
+                    force_uniform_decode = False
```
Comment (vadiklyutiy) [inline vllm/v1/worker/gpu_model_runner.py]: Why this code should be here? 
I think inside `_is_uniform_decode` is better place.

**Comment by vadiklyutiy** [comment]
Thank you for you PR, mixed up spec-decode with prefill is really very fragile thing. 

Could you add a unit test for this case, based on you script in Test Plan?

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (xiatwhu) [reply vllm/v1/worker/gpu_model_runner.py]: > Why this code should be here? I think inside `_is_uniform_decode` is better place.

We use `scheduler_output` to determine if a request is in the prefill phase. If we want to perform this check in `_is_uniform_batch`, we must pass the relevant parameters to `_determine_batch_execution_and_padding` and `_is_uniform_batch` (a static method). The `force_uniform_decode` parameter provided by `_is_uniform_batch` can solve this problem perfectly. `GpuModelRunner._dummy_run` also uses these parameters to distinguish between `mixed_batch` and `uniform_batch`.

**Comment by avesed** [comment]
Independent confirmation of this bug and root cause, with additional scope evidence — full write-up in #49918 (filed to document this PR's bug; it had no linked issue).

We hit the same corruption on Qwen3.6 GDN hybrids and root-caused it independently with capture-vs-replay instrumentation before finding this PR. Confirming your diagnosis, plus some data points that may help land this:

- **Mechanism detail**: the misclassified prefill replays the captured FULL spec-verify graph while the builder produced *prefill* metadata, so the spec persistent-buffer refresh never runs → the captured kernels read capture-time NULL(0) state indices → the NULL guard in `fused_sigmoid_gating_delta_rule_update` (and conv update) skips the recurrent-state **write** entirely. The prompt's state is never created — pre-verify state reads are all-zero at the correct slots (measured), which is why output is deterministic garbage rather than noise.
- **Proposer-independent**: reproduces with `method=ngram` with no draft head loaded — it's purely prefill-time dispatch.
- **Quant-independent**: W4A16 and W8A8 both reproduce; **still present in `vllm/vllm-openai:v0.26.0`** (and v0.25.1).
- **Perfect per-prompt predictor**: each of 6 prompts garbles at exactly `K = prompt_tokens − 1` and is clean at every other K ∈ [2,15]; N equal-length (K+1)-token prompts batched garble too.
- **Dense models are immune** (verified with dense Qwen3-8B + ngram): the mis-dispatch still happens but spec-verify attention over a fresh sequence is mathematically identical to prefill attention — only recurrent-state targets corrupt. So gating on hybrid, as this PR does, matches where the damage is.

We also validated an equivalent minimal guard on a v0.25.1 fork (AND `uniform_decode` with `(num_computed_tokens_cpu >= num_prompt_tokens).all()` over the batch when `force_uniform_decode is None`): all garbage configs became clean, the (K+1)-token prefill dispatches PIECEWISE non-uniform, and genuine verify decode keeps its FULL uniform graphs with unchanged acceptance length and throughput — supporting that this class of fix has no spec-decode perf cost.

Hope this helps get this merged — it's a silent correctness bug on any GDN/hybrid deployment serving raw completions.

(AI-assisted analysis; details and repro in #49918.)

**Comment by JulianZJN** [comment]
Independent data point for this fix, from chasing #55766 (all-NaN logits on Qwen3.8-27B with ngram spec decode + prefix caching).

The misclassification is not limited to prompts that are `K+1` tokens long. With `mamba_cache_mode=align` the prefill is split at the Mamba block boundary, so any prompt with `len % block_size == K+1` ends in a `K+1`-token prefill chunk that is dispatched to the FULL spec-verify graph (block size 816, `K=5`: every prompt of `N*816 + 6` tokens). And the consequence gets worse once the server has served a spec-decode step: the GDN builder doesn't refresh its persistent spec buffers on that step (it sees a prefill), so the replay writes conv/SSM state to the *previous* request's block ids. In my traces one of those blocks had been reallocated as the running request's first attention block; it turns non-finite in the same forward, the request returns NaN logits, and because it is a full block it is put in the prefix cache and poisons the next request that restores it.

Applied the four `gpu_model_runner.py` hunks of this PR (head `41edfaf7`; the only other file is the unit test) on top of main `71fc70d3` (Qwen3.8-27B BF16, TP4, 4× RTX 4090, ngram `num_speculative_tokens=5`, FULL_AND_PIECEWISE):

- 10 requests of `20*816 + 6` tokens: 5–9 requests with NaN logits before, 0/10 after (two server starts)
- A -> B (prefix-cache restore) -> B with new `cache_salt`, `r = 6`, `N = 20` and `34`, 3 seeds each: 2/3, 2/3, 0/3 before, all 0/3 after
- the colliding step now runs PIECEWISE (417 regular spec steps still replay FULL), and per-block checksums of the layer-0 conv/SSM state on that step (whole block pool) are identical to a `cudagraph_mode=PIECEWISE` run
- at temperature 0, 4 of 5 seeds match the PIECEWISE run token-for-token over 200 tokens; the fifth diverges at token 82 between two near-tied candidates. I did not establish full parity: layer-0 state is equal at that step, which says nothing about the other layers
- `test_compute_force_uniform_decode` fails without the runner change and passes with it

Update on the preemption case I asked about: it is reachable, and the guard does not cover it. A preempted request restarts with `num_computed_tokens = 0` but keeps its output tokens, and its replay is chunked on the block grid, so the last replay chunk is `(num_prompt_tokens + num_output_tokens) % block_size` tokens. On a TP4 run with real preemptions (`vllm:num_preemptions_total` moving, small `--num-gpu-blocks-override`) and these hunks applied, a victim with a `7*816 + 5` token prompt and one generated token replayed like this:

```
step=4512 mode=PIECEWISE sched=[6,6,5] k=[5,5,0] resumed=[victim] num_output_tokens=[811,5,0] num_ctx_requests=1 force_uniform_decode=False
step=4910 mode=FULL      sched=[6,6]   k=[5,0]   resumed=[victim] num_output_tokens=[1599,1]  num_ctx_requests=0 force_uniform_decode=None
```

The first replay step is caught because `num_output_tokens == 0` there; the 6-token tail is not, since `is_context_phase` is `num_output_tokens == 0` and that request already has an output token, so `_compute_force_uniform_decode` returns `None` and the step runs FULL. I did not see corrupted requests in these runs, so this is a gap in what the guard covers, not a second NaN reproduction. One way to cover it is to key the check on whether each request's step is actually a verify step, the same predicate the GDN builder uses to decide what a spec row is (`num_scheduled_tokens == len(draft_tokens) + 1`), so the dispatcher and the builder cannot disagree:

```python
if iteration_details.num_ctx_requests > 0 or any(
    n != 1 + len(scheduler_output.scheduled_spec_decode_tokens.get(rid, ()))
    for rid, n in scheduler_output.num_scheduled_tokens.items()
):
    return False
```

I have not run that variant. The V2 runner moves resumed requests into `scheduled_new_reqs` and is not affected, but ngram speculation runs on the V1 runner.

Everything behind these numbers (drivers, per-request tables, the applied diff with its hash, preemption traces) is in https://github.com/JulianZJN/vllm-55766-evidence.

@njhill @vadiklyutiy, could you take another look at the runner guard with this independent frozen-input A/B in mind? The new-request corruption is reproduced and removed by the current patch; the preemption case above is only a demonstrated coverage gap, not a second corruption result.

**Comment by amittell** [comment]
Independent reproduction of #49918 on current main, and this PR fixes it.

Setup: main `9f07d023d07f`, Model Runner V1, one GB10 (sm_121), Qwen3.8-27B-NVFP4 with its MTP head, `num_speculative_tokens` 4, FULL_AND_PIECEWISE cudagraphs.

- A raw 5-token prompt (exactly `1 + num_speculative_tokens`) was dispatched FULL with `uniform=True` for a 5-token prefill row and came back as garbage 2/2 (`' SpainFranceFrance$/,$/...'`). 4- and 6-token prompts were clean.
- With this PR's runner hunks (`41edfaf7`) merged onto the same commit, the same 5-token step ran PIECEWISE and answered `' Paris.'` 2/2.

The step classification was read with a probe that logs every step where a batch containing a non-verify row is classified as uniform decode. It fires on the 5-token case and stays at 0 over about 8,100 steps of GSM8K traffic, so the trigger is narrow: a prefill whose final chunk is exactly `1 + num_speculative_tokens` tokens. #55766's 09-19 analysis lands on the same mechanism.
